x/dymns/keeper/hooks.go

```diff
@@ -0,0 +1,365 @@
+
+var _ epochstypes.EpochHooks = epochHooks{}
+
+type epochHooks struct {
+	Keeper
+}
+
+// GetEpochHooks returns the epoch hooks for the module.
+func (k Keeper) GetEpochHooks() epochstypes.EpochHooks {
+	return epochHooks{
+		Keeper: k,
+	}
+}
+
+// BeforeEpochStart is the epoch start hook.
+func (e epochHooks) BeforeEpochStart(_ sdk.Context, _ string, _ int64) error {
+	return nil
+}
+
+// AfterEpochEnd is the epoch end hook.
+func (e epochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	miscParams := e.MiscParams(ctx)
+
+	if epochIdentifier != miscParams.EndEpochHookIdentifier {
+		return nil
+	}
+
+	logger := e.Logger(ctx).With("hook", "After-Epoch-End", "epoch-number", epochNumber, "epoch-identifier", epochIdentifier)
+
+	if miscParams.EnableTradingName {
+		if err := e.processActiveDymNameSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	if miscParams.EnableTradingAlias {
+		if err := e.processActiveAliasSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// processActiveDymNameSellOrders process the finished Dym-Name Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Dym-Name ownership will be transferred.
+func (e epochHooks) processActiveDymNameSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeName, logger)
+
+	if len(finishedSOs) < 1 || len(aSoe.Records) < 1 {
+		return nil
+	}
+
+	logger.Info("processing finished SOs", "count", len(finishedSOs))
+
+	for _, so := range finishedSOs {
+		// each order should be processed in a branched context, if error, discard the state change
+		// and process next order, to prevent chain reaction when an individual order failed to process
+		errApplyStateChange := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			if so.HighestBid == nil {
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeName)
+				return nil
+			}
+
+			if err := e.CompleteDymNameSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
+			}
+
+			return nil
+		})
+
+		if errApplyStateChange == nil {
+			aSoe.Remove(so.AssetId)
+		}
+	}
+
+	if err := e.SetActiveSellOrdersExpiration(ctx, aSoe, dymnstypes.TypeName); err != nil {
+		e.Keeper.Logger(ctx).Error("failed to update active SO expiry", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// processActiveAliasSellOrders process the finished Alias Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Alias linking will be updated.
+func (e epochHooks) processActiveAliasSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeAlias, logger)
+
+	if len(finishedSOs) < 1 || len(aSoe.Records) < 1 {
+		return nil
+	}
+
+	logger.Info("processing finished SOs", "count", len(finishedSOs))
+
```

Comment:
why `aSoe`? (I guess `activeSellOrdersExpiration`? not clear) 
it says we only process the finished sell-order so why we have 2 variables we check for (both `finishedSos` and `aSoe`)?

Reply:
`aSoe` is the list contains pointer to the order, and the `finishedSos` is the SO list fetched by using the pointers.